### PR TITLE
bpo-36073: Raise ProgrammingError on recursive usage of cursors in sqlite converters

### PR DESCRIPTION
**Pull request by sir-sigurd**
https://bugs.python.org/issue36073

### CODE REVIEW

**Review by auvipy — APPROVED** [review]

**Review by auvipy — APPROVED** [review]

File: Lib/sqlite3/test/regression.py [hunk 1]
```diff
@@ -444,11 +444,46 @@ class UnhashableType(type):
             self.con.execute('SELECT %s()' % aggr_name)
 
 
+class ConverterProggrammingErrorTestCase(unittest.TestCase):
```
Comment (BoboTiG) [inline Lib/sqlite3/test/regression.py]: Small typo:
```suggestion
class ConverterProgrammingErrorTestCase(unittest.TestCase):
```

**Comment by BoboTiG** [comment]
Thanks @sir-sigurd, I think a NEWs entry would be interesting, don't you?

**Comment by erlend-aasland** [comment]
@sir-sigurd, can you rebase onto master?

File: Modules/_sqlite/cursor.c [hunk 2]
```diff
@@ -349,12 +362,8 @@ static int check_cursor(pysqlite_Cursor* cur)
...
     }
 
-    if (cur->locked) {
-        PyErr_SetString(pysqlite_ProgrammingError, "Recursive use of cursors not allowed.");
-        return 0;
-    }
-
-    return pysqlite_check_thread(cur->connection) && pysqlite_check_connection(cur->connection);
+    return check_cursor_locked(cur) &&
+        pysqlite_check_thread(cur->connection) && pysqlite_check_connection(cur->connection);
```
Comment (erlend-aasland) [inline Modules/_sqlite/cursor.c]: Please format these lines according to PEP 7.

File: Modules/_sqlite/cursor.c [hunk 3]
```diff
@@ -29,10 +29,23 @@ PyObject* pysqlite_cursor_iternext(pysqlite_Cursor* self);
 
 static const char errmsg_fetch_across_rollback[] = "Cursor needed to be reset because of commit/rollback and can no longer be fetched from.";
 
+static int check_cursor_locked(pysqlite_Cursor* cur)
+{
+    if (cur->locked) {
+        PyErr_SetString(pysqlite_ProgrammingError, "Recursive use of cursors not allowed.");
```
Comment (erlend-aasland) [inline Modules/_sqlite/cursor.c]: Line is longer than 79 characters, ref. PEP 7.

File: Modules/_sqlite/cursor.c [hunk 4]
```diff
@@ -29,10 +29,23 @@ PyObject* pysqlite_cursor_iternext(pysqlite_Cursor* self);
 
 static const char errmsg_fetch_across_rollback[] = "Cursor needed to be reset because of commit/rollback and can no longer be fetched from.";
 
+static int check_cursor_locked(pysqlite_Cursor* cur)
```
Comment (erlend-aasland) [inline Modules/_sqlite/cursor.c]: PEP 7

```suggestion
static int
check_cursor_locked(pysqlite_Cursor *cur)
```

File: Misc/NEWS.d/next/Library/2019-06-22-11-01-45.bpo-36073.ED8mB9.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+Raise ProgrammingError on recursive usage of cursors in sqlite converters.
```
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Library/2019-06-22-11-01-45.bpo-36073.ED8mB9.rst]: Missing reST formatting and attribution.

```suggestion
Raise :exc:`~sqlite3.ProgrammingError` on recursive usage of cursors in
:mod:`sqlite3` converters. Patch by Sergey Fedoseev.
```

File: Lib/sqlite3/test/regression.py [hunk 6]
```diff
@@ -399,11 +399,45 @@ def log(self, *args):
...
+        self.cur.execute('select x as "x [CURSOR_ITER]", x from test')
+        with self.assertRaises(sqlite.ProgrammingError):
+            self.cur.fetchone()
+
 
 def suite():
     regression_suite = unittest.makeSuite(RegressionTests, "Check")
     return unittest.TestSuite((
         regression_suite,
+        unittest.makeSuite(ConverterProgrammingErrorTestCase),
```
Comment (erlend-aasland) [inline Lib/sqlite3/test/regression.py]: This needs to be rebased onto master in order to fit in with the changes in the test suite.

**Review by erlend-aasland — CHANGES_REQUESTED** [review]

**Comment by sir-sigurd** [comment]
@erlend-aasland 
I'll try to rebase in the near future, though I almost do not remember anything about this.

**Comment by erlend-aasland** [comment]
> I'll try to rebase in the near future, though I almost do not remember anything about this.

Thanks, there's no hurry :)

**Review by erlend-aasland — APPROVED** [review]
LGTM!

**Comment by erlend-aasland** [comment]
@serhiy-storchaka would you mind reviewing this?

**Comment by sir-sigurd** [comment]
Closed in favor of GH-29054.

**Pull request closed without merging**